bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/servlet/CameraServlet.java

```diff
@@ -0,0 +1,311 @@
+                                Ffmpeg localMjpeg = handler.ffmpegMjpeg;
+                                if (localMjpeg != null) {
+                                    localMjpeg.stopConverting();
+                                }
+                            } else {
+                                handler.closeChannel(handler.getTinyUrl(handler.mjpegUri));
+                            }
+                            logger.debug("All ipcamera.mjpeg streams have stopped.");
+                        }
+                        return;
+                    }
+                } while (true);
+            case "/autofps.mjpeg":
+                req.getSession().setMaxInactiveInterval(0);
+                autofpsStreamsOpen++;
+                handler.streamingAutoFps = true;
+                output = new StreamOutput(resp);
+                int counter = 0;
+                do {
+                    try {
+                        if (handler.motionDetected) {
+                            output.sendSnapshotBasedFrame(handler.getSnapshot());
+                        } else if (counter % 8 == 0) {// every 8 seconds if no motion
+                            output.sendSnapshotBasedFrame(handler.getSnapshot());
+                        }
+                        counter++;
+                        Thread.sleep(1000);
+                    } catch (InterruptedException | IOException e) {
+                        // Never stop streaming until IOException. Occurs when browser stops the stream.
+                        autofpsStreamsOpen--;
+                        if (autofpsStreamsOpen == 0) {
+                            handler.streamingAutoFps = false;
+                            logger.debug("All autofps.mjpeg streams have stopped.");
+                        }
+                        return;
+                    }
+                } while (true);
+            case "/instar":
+                InstarHandler instar = new InstarHandler(handler);
+                instar.alarmTriggered(pathInfo + "?" + req.getQueryString());
+                return;
+            case "/ipcamera0.ts":
+            default:
+                if (pathInfo.contains(".ts")) {
+                    sendFile(resp, pathInfo, "video/MP2T");
+                } else if (pathInfo.contains(".gif")) {
+                    sendFile(resp, pathInfo, "image/gif");
+                } else if (pathInfo.contains(".jpg")) {
+                    // Allow access to the preroll and postroll jpg files
+                    sendFile(resp, pathInfo, "image/jpg");
+                } else if (pathInfo.contains(".m4s") || pathInfo.contains(".mp4")) {
+                    sendFile(resp, pathInfo, "video/mp4");
+                }
+                return;
+        }
+    }
+
+    private void sendFile(HttpServletResponse response, @Nullable String fileUri, String contentType)
+            throws IOException {
+        File file = new File(handler.cameraConfig.getFfmpegOutput() + fileUri);
```

Comment:
What happens if the user enters some `../../` in the path. Can he/she read the whole filesystem?

Reply:
No only the contents of 1 folder and it will not go into sub folders. Even then it only serves a few file extensions like *.ts and *.mp4 which are necessary for HLS and any recordings done by the binding. The rest must be exact matches. So only media files and a call to fetch password.txt in that folder would fail.

Reply:
If I see correctly, it will serve any file in the whole filesystem which name contains one of the extensions in the code. What would prevent the code to go into subfolders or parent directories via `../`?

Reply:
Thank you for raising this, as I was not aware you could mix ABSOLUTE and RELATIVE stuff like that. It appears to not be an issue when tested in 3 browsers. Safari, Edge and Chrome were resolved by the browser before the GET request is made, could another browser handle it differently? I also tried wget and curl on the linux terminal and they both also failed to act differently.

In all 5 cases I tested, entering `http://openhab:8080/ipcamera/mycam/../ipcamera51.ts` into the address bar simply resolves to a get request being sent to `http://openhab:8080/ipcamera/ipcamera51.ts`
The servlet is not listening to requests at that url, so it gets a generic openhab `Requested content not found`.
I guess it will return a file IF someone had a file named as that in the static folder, but in that case you can already request the file in the html/static folder as that is its intention.

The issue I see is that with low latency HLS you can be making file requests every 250ms or even faster depending on how far you wish to push things in the search for low latency video feeds. So I would be happy to insert more code on the other file formats, but I am reluctant to bloat the calls to the .ts format for this reason unless you advise it should be done?

One thought would be to..

String truncated = pathInfo.substring(pathInfo.lastIndexOf("/"));

This way it throws away anything before the last "/" in the requested URL, and will not go anywhere other than a single folder which is fine with the binding. Is the overhead cost worth doing this? If yes then I am happy to do it I just need someone with more knowledge than me to make the call.

a) Do it to all but .ts files.
b) Do to all files.
c) its not an issue, leave as is.

Reply:
Better not rely on the client resolving the URL to a safe URL. I don't know if Jetty has some internal mechanism to resolve the URL before it is handed over to your servlet. I'd go for the substring call to limit the path to a file instead of an entire path. I don't see any performance issues invoking this operation. I'd apply it in `sendFile()` to be sure future enhancements don't open some door.

Reply:
DONE. It is changed so it should not be possible to reach parent or child folders.